### PR TITLE
[runtime] Fix a bunch of warnings.

### PR DESCRIPTION
**Pull request by alexrp**

### CODE REVIEW

File: mono/metadata/monitor.c [hunk 1]
```diff
@@ -333,7 +333,7 @@ mono_locks_dump (gboolean include_untaken)
 					to_recycle++;
 			} else {
 				if (!monitor_is_on_freelist ((MonoThreadsSync *)mon->data)) {
-					MonoObject *holder = (MonoObject *)mono_gchandle_get_target ((guint32)mon->data);
+					MonoObject *holder = (MonoObject *)mono_gchandle_get_target ((guint32)(size_t)mon->data);
```
Comment (lewurm) [inline mono/metadata/monitor.c]: I'm not sure what's the story of the `g*` types, but I think we shouldn't mix them at least. So I would suggest `gsize` instead.

File: mono/mini/mini-amd64.h [hunk 2]
```diff
@@ -220,28 +220,31 @@ typedef struct MonoCompileArch {
 
 #ifdef TARGET_WIN32
 
-static AMD64_Reg_No param_regs [] = { AMD64_RCX, AMD64_RDX, AMD64_R8, AMD64_R9 };
+static G_GNUC_UNUSED AMD64_Reg_No param_regs [] = { AMD64_RCX, AMD64_RDX, AMD64_R8, AMD64_R9 };
```
Comment (lewurm) [inline mono/mini/mini-amd64.h]: hm, that's ugly...  can't we move them somewhere else and declare them as `external` here?

Comment (alexrp) [inline mono/mini/mini-amd64.h]: Then declaring them becomes rather cumbersome since you have to explicitly specify the array size.

Comment (lewurm) [inline mono/mini/mini-amd64.h]: I see :/

File: mono/metadata/threads.c [hunk 3]
```diff
@@ -5198,6 +5191,13 @@ mono_threads_detach_coop (gpointer cookie, gpointer *dummy)
 }
 
 #if 0
+static int
```
Comment (luhenry) [inline mono/metadata/threads.c]: @kumpera shouldn't we just remove this code if it's unused?

File: mono/metadata/sre-encode.c [hunk 4]
```diff
@@ -36,7 +36,7 @@ static guint32 create_typespec (MonoDynamicImage *assembly, MonoType *type);
 static void    encode_type (MonoDynamicImage *assembly, MonoType *type, SigBuffer *buf);
 static guint32 mono_image_typedef_or_ref (MonoDynamicImage *assembly, MonoType *type);
 
-static guint32
+static G_GNUC_UNUSED guint32
```
Comment (luhenry) [inline mono/metadata/sre-encode.c]: Can you just put the whole function between `#if G_BYTE_ORDER != G_LITTLE_ENDIAN` ?

File: mono/metadata/sre-encode.c [hunk 5]
```diff
@@ -476,7 +476,7 @@ mono_dynimage_encode_locals (MonoDynamicImage *assembly, MonoReflectionILGen *il
  * Copy len * nelem bytes from val to dest, swapping bytes to LE if necessary.
  * dest may be misaligned.
  */
-static void
+static G_GNUC_UNUSED void
```
Comment (luhenry) [inline mono/metadata/sre-encode.c]: Can you just put the whole function between `#if G_BYTE_ORDER != G_LITTLE_ENDIAN` ?

File: mono/metadata/sre-save.c [hunk 6]
```diff
@@ -109,7 +109,7 @@ find_index_in_table (MonoDynamicImage *assembly, int table_idx, int col, guint32
  * Copy len * nelem bytes from val to dest, swapping bytes to LE if necessary.
  * dest may be misaligned.
  */
-static void
+static G_GNUC_UNUSED void
```
Comment (luhenry) [inline mono/metadata/sre-save.c]: Can you just put the whole function between #if G_BYTE_ORDER != G_LITTLE_ENDIAN ?

File: mono/metadata/monitor.c [hunk 7]
```diff
@@ -333,7 +333,7 @@ mono_locks_dump (gboolean include_untaken)
 					to_recycle++;
 			} else {
 				if (!monitor_is_on_freelist ((MonoThreadsSync *)mon->data)) {
-					MonoObject *holder = (MonoObject *)mono_gchandle_get_target ((guint32)mon->data);
+					MonoObject *holder = (MonoObject *)mono_gchandle_get_target ((guint32)(gsize)mon->data);
```
Comment (luhenry) [inline mono/metadata/monitor.c]: Please use `GPOINTER_TO_UINT` instead of `(guint32)(gsize)`

File: mono/metadata/monitor.c [hunk 8]
```diff
@@ -390,7 +390,7 @@ mon_new (gsize id)
 		new_ = NULL;
 		for (marray = monitor_allocated; marray; marray = marray->next) {
 			for (i = 0; i < marray->num_monitors; ++i) {
-				if (mono_gchandle_get_target ((guint32)marray->monitors [i].data) == NULL) {
+				if (mono_gchandle_get_target ((guint32)(gsize)marray->monitors [i].data) == NULL) {
```
Comment (luhenry) [inline mono/metadata/monitor.c]: Please use `GPOINTER_TO_UINT` instead of `(guint32)(gsize)`

File: mono/metadata/monitor.c [hunk 9]
```diff
@@ -400,7 +400,7 @@ mon_new (gsize id)
 							new_->wait_list = g_slist_remove (new_->wait_list, new_->wait_list->data);
 						}
 					}
-					mono_gchandle_free ((guint32)new_->data);
+					mono_gchandle_free ((guint32)(gsize)new_->data);
```
Comment (luhenry) [inline mono/metadata/monitor.c]: Please use `GPOINTER_TO_UINT` instead of `(guint32)(gsize)`

File: mono/metadata/monitor.c [hunk 10]
```diff
@@ -457,7 +457,7 @@ alloc_mon (MonoObject *obj, gint32 id)
 
 	mono_monitor_allocator_lock ();
 	mon = mon_new (id);
-	mon->data = (void *)(size_t)mono_gchandle_new_weakref (obj, TRUE);
+	mon->data = (gpointer)(gsize)mono_gchandle_new_weakref (obj, TRUE);
```
Comment (luhenry) [inline mono/metadata/monitor.c]: Please use `GUINT_TO_POINTER` instead of `(gpointer)(gsize)`

File: mono/metadata/monitor.c [hunk 11]
```diff
@@ -468,7 +468,7 @@ static void
 discard_mon (MonoThreadsSync *mon)
 {
 	mono_monitor_allocator_lock ();
-	mono_gchandle_free ((guint32)mon->data);
+	mono_gchandle_free ((guint32)(gsize)mon->data);
```
Comment (luhenry) [inline mono/metadata/monitor.c]: Please use GPOINTER_TO_UINT instead of (guint32)(gsize)

File: mono/metadata/monitor.c [hunk 12]
```diff
@@ -1118,7 +1118,7 @@ mono_monitor_get_object_monitor_gchandle (MonoObject *object)
 
 	if (lock_word_is_inflated (lw)) {
 		MonoThreadsSync *mon = lock_word_get_inflated_lock (lw);
-		return (guint32)mon->data;
+		return (guint32)(gsize)mon->data;
```
Comment (luhenry) [inline mono/metadata/monitor.c]: Please use GPOINTER_TO_UINT instead of (guint32)(gsize)

File: mono/mini/mini-amd64-gsharedvt.c [hunk 13]
```diff
@@ -57,7 +57,7 @@ storage_name (ArgStorage st)
 	}
 }
 
-static char *
+static G_GNUC_UNUSED char *
```
Comment (luhenry) [inline mono/mini/mini-amd64-gsharedvt.c]: Please put it between `#ifdef DEBUG_AMD64_GSHAREDVT`

File: mono/mini/mini-amd64-gsharedvt.h [hunk 14]
```diff
@@ -37,7 +37,7 @@ typedef enum {
 	GSHAREDVT_RET_NUM,
 } GSharedVtRetMarshal;
 
-static const char* ret_marshal_name[] = {
+static G_GNUC_UNUSED const char* ret_marshal_name[] = {
```
Comment (luhenry) [inline mono/mini/mini-amd64-gsharedvt.h]: Please put it between #ifdef DEBUG_AMD64_GSHAREDVT

**Pull request closed without merging**